ext/standard/mail.c

```diff
@@ -96,20 +97,210 @@ PHP_FUNCTION(ezmlm_hash)
+
+	/* https://tools.ietf.org/html/rfc2822#section-2.2.3 */
+	while (len < value->len) {
+		if (*value->val == '\r') {
+			if (value->len - len > 4
+				&&  *value->val+1 == '\n'
+				&& (*value->val+2 == ' '  || *value->val+2 == '\t')
+				&& (*value->val+3 >= 32   && *value->val+3 <= 126)) {
+				len += 4;
+				continue;
+			}
+			return FAILURE;
+		}
+		if (*value->val < 33 || *value->val > 126) {
+			return FAILURE;
+		}
+		len++;
+	}
+	return SUCCESS;
+}
+
+
+static zend_bool php_mail_build_headers_check_field_name(zend_string *key)
+{
+	size_t len = 0;
+
+	/* https://tools.ietf.org/html/rfc2822#section-2.2 */
+	while (len < key->len) {
+		if (*key->val < 33 || *key->val > 126 || *key->val == ':') {
+			return FAILURE;
+		}
+		len++;
+	}
+	return SUCCESS;
+}
+
+
+static void php_mail_build_headers_elems(smart_str *s, zend_string *key, zval *val);
+
+static void php_mail_build_headers_elem(smart_str *s, zend_string *key, zval *val)
+{
+	switch(Z_TYPE_P(val)) {
+		case IS_STRING:
+			if (php_mail_build_headers_check_field_name(key) == FAILURE) {
+				php_error_docref(NULL, E_WARNING, "Header field name (%s) contains invalid chars", ZSTR_VAL(key));
+				return;
+			}
+			if (php_mail_build_headers_check_field_value(val) == FAILURE) {
+				php_error_docref(NULL, E_WARNING, "Header field value (%s) contains invalid chars or format", Z_STRVAL_P(val));
+				return;
+			}
+			smart_str_append(s, key); // Ignore chars after null
+			smart_str_appendl(s, ": ", 2);
+			smart_str_appends(s, Z_STRVAL_P(val)); // Ignore chars after null.
+			smart_str_appendl(s, "\r\n", 2);
+			break;
+		case IS_ARRAY:
+			php_mail_build_headers_elems(s, key, val);
+			break;
+		default:
+			php_error_docref(NULL, E_WARNING, "headers array elements must be string or array (%s)", ZSTR_VAL(key));
+	}
+}
+
+
+static void php_mail_build_headers_elems(smart_str *s, zend_string *key, zval *val)
+{
+	zend_ulong idx;
+	zend_string *tmp_key;
+	zval *tmp_val;
+
+	(void)(idx);
+	ZEND_HASH_FOREACH_KEY_VAL(HASH_OF(val), idx, tmp_key, tmp_val) {
+		if (tmp_key) {
+			php_error_docref(NULL, E_WARNING, "Multiple header key must be numeric index (%s)", ZSTR_VAL(tmp_key));
+			continue;
+		}
+		if (Z_TYPE_P(tmp_val) != IS_STRING) {
+			php_error_docref(NULL, E_WARNING, "Multiple header values must be string (%s)", ZSTR_VAL(key));
+			continue;
+		}
+		php_mail_build_headers_elem(s, key, tmp_val);
+	} ZEND_HASH_FOREACH_END();
+}
+
+
+PHPAPI zend_string *php_mail_build_headers(zval *headers)
+{
+	zend_ulong idx;
+	zend_string *key;
+	zval *val;
+	smart_str s = {0};
+
+	ZEND_ASSERT(Z_TYPE_P(headers) == IS_ARRAY);
+
+	ZEND_HASH_FOREACH_KEY_VAL(HASH_OF(headers), idx, key, val) {
+		if (!key) {
+			php_error_docref(NULL, E_WARNING, "Found numeric header (" ZEND_LONG_FMT ")", idx);
+			continue;
+		}
+		/* https://tools.ietf.org/html/rfc2822#section-3.6 */
+		switch(ZSTR_LEN(key)) {
+			case sizeof("orig-date")-1:
+				if (!strncasecmp("orig-date", ZSTR_VAL(key), ZSTR_LEN(key))) {
```

Comment:
This set of forbidden headers will most certainly break existing libraries. See e.g. [here](https://github.com/swiftmailer/swiftmailer/blob/5.x/lib/classes/Swift/Transport/MailTransport.php#L125).